optimizely/event/dispatcher.go

```diff
@@ -54,10 +58,83 @@ func (*HTTPEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
 		if resp.StatusCode == 204 {
 			success = true
 		} else {
-			fmt.Printf("http.Post invalid response %d", resp.StatusCode)
+			dispatcherLogger.Error(fmt.Sprintf("http.Post invalid response %d", resp.StatusCode), nil)
 			success = false
 		}
 	}
 	callback(success)
+}
+
+// QueueEventDispatcher is a queued version of the event dispatcher that queues, returns success, and dispatches events in the background
+type QueueEventDispatcher struct {
+	eventQueue     Queue
+	eventFlushLock sync.Mutex
+	dispatcher     *HTTPEventDispatcher
+}
+
+// DispatchEvent queues event with callback and calls flush in a go routine.
+func (ed *QueueEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
+
+	ed.eventQueue.Add(event)
+	callback(true)
+	go func() {
+		ed.flushEvents()
+	}()
+}
+
+// flush the events
+func (ed *QueueEventDispatcher) flushEvents() {
+
+	ed.eventFlushLock.Lock()
+
+	defer func(){
+		ed.eventFlushLock.Unlock()
+	}()
+
+	retryCount := 0
+
+	for ed.eventQueue.Size() > 0 {
+		if retryCount > maxRetries {
+			dispatcherLogger.Error(fmt.Sprintf("event failed to send %d times. It will retry on next event sent", maxRetries), nil)
+			break
+		}
+
+		items := ed.eventQueue.Get(1)
+		if len(items) == 0 {
+			// something happened.  Just continue and you should expect size to be zero.
+			continue
+		}
+		event, ok := items[0].(LogEvent)
+		if !ok {
+			// remove it
+			dispatcherLogger.Error("invalid type passed to event dispatcher", nil)
+			ed.eventQueue.Remove(1)
+			continue
+		}
+
+		ed.dispatcher.DispatchEvent(event, func(success bool) {
+			if success {
+				dispatcherLogger.Debug("event sent")
```

Comment:
As mentioned offline, please add more context to the log message like this: 
`dispatcherLogger.Debug(fmt.Sprintf("Dispatched log event %+v", event))`